hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/hudi/HoodieCreateRecordUtils.scala

```diff
@@ -0,0 +1,285 @@
+
+/** Utility class for converting dataframe into RDD[HoodieRecord]. */
+object HoodieCreateRecordUtils {
+  private val log = LoggerFactory.getLogger(getClass)
+
+  def createHoodieRecordRdd(df: DataFrame,
+                            config: HoodieWriteConfig,
+                            parameters: Map[String, String],
+                            recordName: String,
+                            recordNameSpace: String,
+                            writerSchema: Schema,
+                            dataFileSchema: Schema,
+                            operation: WriteOperationType,
+                            instantTime: String,
+                            isPrepped: Boolean) = {
+    val shouldDropPartitionColumns = config.getBoolean(DataSourceWriteOptions.DROP_PARTITION_COLUMNS)
+    val recordType = config.getRecordMerger.getRecordType
+    val autoGenerateRecordKeys: Boolean = !parameters.containsKey(KeyGeneratorOptions.RECORDKEY_FIELD_NAME.key());
+
+    val shouldCombine = if (!isPrepped && WriteOperationType.isInsert(operation)) {
+      parameters(INSERT_DROP_DUPS.key()).toBoolean ||
+        parameters.getOrElse(
+          HoodieWriteConfig.COMBINE_BEFORE_INSERT.key(),
+          HoodieWriteConfig.COMBINE_BEFORE_INSERT.defaultValue()
+        ).toBoolean
+    } else if (!isPrepped && WriteOperationType.isUpsert(operation)) {
+      parameters.getOrElse(
+        HoodieWriteConfig.COMBINE_BEFORE_UPSERT.key(),
+        HoodieWriteConfig.COMBINE_BEFORE_UPSERT.defaultValue()
+      ).toBoolean
+    } else {
+      !isPrepped
+    }
+
+    // NOTE: Avro's [[Schema]] can't be effectively serialized by JVM native serialization framework
+    //       (due to containing cyclic refs), therefore we have to convert it to string before
+    //       passing onto the Executor
+    val dataFileSchemaStr = dataFileSchema.toString
+
+    log.debug(s"Creating HoodieRecords (as $recordType)")
+
+    recordType match {
+      case HoodieRecord.HoodieRecordType.AVRO =>
+        // avroRecords will contain meta fields when isPrepped is true.
+        val avroRecords: RDD[GenericRecord] = HoodieSparkUtils.createRdd(df, recordName, recordNameSpace,
+          Some(writerSchema))
+
+        if (isPrepped) {
+          // For prepped records, check the first record to make sure it has meta fields set.
+          validateMetaFieldsInAvroRecords(avroRecords.first())
+        }
+
+        avroRecords.mapPartitions(it => {
+          val sparkPartitionId = TaskContext.getPartitionId()
+          val keyGenProps = new TypedProperties(config.getProps)
+          if (autoGenerateRecordKeys) {
+            keyGenProps.setProperty(KeyGenUtils.RECORD_KEY_GEN_PARTITION_ID_CONFIG, String.valueOf(sparkPartitionId))
+            keyGenProps.setProperty(KeyGenUtils.RECORD_KEY_GEN_INSTANT_TIME_CONFIG, instantTime)
+          }
+          val keyGenerator = HoodieSparkKeyGeneratorFactory.createKeyGenerator(keyGenProps)
```

Comment:
can we make this optional as well. for prepped path, it should be empty.

Reply:
Done.